File: structural/typescript/bridge_structured/abstractions.ts
import { DrawingAPI, Device } from './interfaces';

export abstract class Shape {
  protected drawingAPI: DrawingAPI;
  constructor(api: DrawingAPI) { this.drawingAPI = api; }
  abstract draw(): void;
  abstract resize(percentage: number): void;
}

export class Circle extends Shape {
  private x: number;
  private y: number;
  private radius: number;
  
  constructor(x: number, y: number, radius: number, api: DrawingAPI) {
    super(api);
    this.x = x; this.y = y; this.radius = radius;
  }
  draw(): void { this.drawingAPI.drawCircle(this.x, this.y, this.radius); }
  resize(percentage: number): void {
    this.radius *= percentage / 100;
    console.log(`Resizing circle to ${this.radius} radius`);
  }
}

export class Rectangle extends Shape {
  private x: number;
  private y: number;
  private width: number;
  private height: number;
  
  constructor(x: number, y: number, width: number, height: number, api: DrawingAPI) {
    super(api);
    this.x = x; this.y = y; this.width = width; this.height = height;
  }
  draw(): void { this.drawingAPI.drawRectangle(this.x, this.y, this.width, this.height); }
  resize(percentage: number): void {
    this.width *= percentage / 100;
    this.height *= percentage / 100;
    console.log(`Resizing rectangle to ${this.width}x${this.height}`);
  }
}

export class RemoteControl {
  protected device: Device;
  constructor(device: Device) { this.device = device; }
  
  togglePower(): void {
    if (this.device.getVolume() >= 0) { this.device.turnOff(); }
    else { this.device.turnOn(); }
  }
  volumeUp(): void { this.device.setVolume(this.device.getVolume() + 1); }
  volumeDown(): void { this.device.setVolume(this.device.getVolume() - 1); }
}

export class AdvancedRemoteControl extends RemoteControl {
  mute(): void {
    console.log('Muting device');
    this.device.setVolume(0);
  }
}

File: structural/dart/bridge_structured/abstractions.dart
import 'interfaces.dart';

abstract class Shape {
  final DrawingAPI drawingAPI;
  
  Shape(this.drawingAPI);
  
  void draw();
  void resize(double percentage);
}

class Circle extends Shape {
  double _x, _y, _radius;
  
  Circle(this._x, this._y, this._radius, DrawingAPI api) : super(api);
  
  @override
  void draw() {
    drawingAPI.drawCircle(_x, _y, _radius);
  }
  
  @override
  void resize(double percentage) {
    _radius *= percentage / 100;
    print('Resizing circle to $_radius radius');
  }
}

class Rectangle extends Shape {
  double _x, _y, _width, _height;
  
  Rectangle(this._x, this._y, this._width, this._height, DrawingAPI api) : super(api);
  
  @override
  void draw() {
    drawingAPI.drawRectangle(_x, _y, _width, _height);
  }
  
  @override
  void resize(double percentage) {
    _width *= percentage / 100;
    _height *= percentage / 100;
    print('Resizing rectangle to ${_width}x$_height');
  }
}

class RemoteControl {
  final Device _device;
  
  RemoteControl(this._device);
  
  void togglePower() {
    if (_device.getVolume() >= 0) {
      _device.turnOff();
    } else {
      _device.turnOn();
    }
  }
  
  void volumeUp() {
    int currentVolume = _device.getVolume();
    _device.setVolume(currentVolume + 1);
  }
  
  void volumeDown() {
    int currentVolume = _device.getVolume();
    _device.setVolume(currentVolume - 1);
  }
}

class AdvancedRemoteControl extends RemoteControl {
  AdvancedRemoteControl(Device device) : super(device);
  
  void mute() {
    print('Muting device');
    _device.setVolume(0);
  }
}
